fix(core): handle synthetic props in Directive host bindings correctly

AndrewKushnir opened the pull request:
Prior to this change, animations-related runtime logic assumed that the `@HostBinding` and `@HostListener` with synthetic (animations) props are used for Components only. However having `@HostBinding` and `@HostListener` with synthetic props on Directives is supported by View Engine. This commit updates the logic to select correct renderer to execute instructions (current renderer for Directives and sub-component renderer for Components).

This PR resolves #35501.

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
fix(core): handle synthetic props in Directive host bindings correctly

Prior to this change, animations-related runtime logic assumed that the @HostBinding and @HostListener with synthetic (animations) props are used for Components only. However having @HostBinding and @HostListener with synthetic props on Directives is also supported by View Engine. This commit updates the logic to select correct renderer to execute instructions (current renderer for Directives and sub-component renderer for Components).

This PR resolves #35501.

kara commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1263,11 +1263,13 @@ function invokeDirectivesHostBindings(tView: TView, lView: LView, tNode: TNode)
+    for (let dirIndex = start; dirIndex < end; dirIndex++) {
+      const def = tView.data[dirIndex] as DirectiveDef<any>;
+      const directive = lView[dirIndex];
+      setCurrentDirectiveIndex(dirIndex);
```
Hmm this means that every host binding, regardless of whether there are any synthetic host bindings, will pay the cost of keeping track of the directive index global, right? If I read the comment right, we can't generate something different at compile-time because we don't know if a directive might be extended by a component....

Can we run benchmarks on this change to see impact on apps with lots of host bindings? If we don't have a micro-benchmark for host bindings, we should create one.

AndrewKushnir replied:
Thanks for the review @kara. We don't seem to have micro benchmarks for host bindings, so I created one (in a separate PR): https://github.com/angular/angular/pull/35705. Once that PR lands, I will measure perf impact of these changes. Thank you.

kara commented on packages/core/src/render3/instructions/host_property.ts:
```diff
@@ -68,8 +70,17 @@ export function ɵɵupdateSyntheticHostBinding<T>(
+        currentDef === null || isComponentDef(currentDef) ? loadComponentRenderer : null;
     elementPropertyInternal(
-        tView, lView, nodeIndex, propName, value, sanitizer, true, loadComponentRenderer);
+        tView, lView, nodeIndex, propName, value, sanitizer, true, loadComponentRendererFn);
```
If there is a component with a directive on the same node, and the directive has a synthetic host binding, I take it that the correct renderer is *not* the component renderer? Is this why we can't just fall back to `lView[RENDERER]` if the component view isn't found?  Can we add a test that showcases this?

mhevery replied:
Discussed with @AndrewKushnir offline
- [ ] We think that `elementPropertyInternal` should not take `loadComponentRendererFn` but instead take the `Renderer` directly. The implication of this is that the renderer lookup logic should move from `elementPropertyInternal` to `ɵɵupdateSyntheticHostBinding`
- [ ] The above will require that instead of passing `nodeIndex` we pass in `TNode`. Implication is that `const nodeIndex = getSelectedIndex();` should change to `const tNode = getSelectedTNode();`

kara reviewed (commented):
Mostly questions

AndrewKushnir commented on packages/core/test/acceptance/host_binding_spec.ts:
```diff
@@ -175,6 +178,266 @@ describe('host bindings', () => {
     });
   });
 
+  describe('with synthetic (animations) props', () => {
```
👍

AndrewKushnir commented:
Hi @kara, I performed performance measurements using newly added benchmark for host bindings. I tested 2 scenarios:

- 1 directive with 1 host binding + 1 host listeners attached to an element. The results (master vs this branch) are almost identical.
- 100 directives with 1 host binding + 1 host listeners attached to an element (to emulate a case that you mentioned in the comment). The result of this change is a bit more noticeable, but still low - around **0.5%** (12.33us vs 12.26us in best cases).

Thank you.

pkozlowski-opensource commented:
I'm putting this PR in the "cleanup" state for a tiny bit since with @AndrewKushnir we've observed some funky numbers when running perf tests. We talked and for 99,99% this is just a case of the rebase that needs to be done, but I will confirm this tomorrow when the rebase is done.

AndrewKushnir commented:
Thanks for looking at the changes in this PR @pkozlowski-opensource! I've rebased this PR, so it should be up-to-date with the current master (the `host_binding` benchmark should be the same now). Thank you.

mhevery commented on packages/core/test/acceptance/host_binding_spec.ts:
```diff
@@ -6,10 +6,13 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {animate, state, style, transition, trigger} from '@angular/animations';
```
nit: `animate` is un-needed import

mhevery commented on packages/core/src/render3/instructions/listener.ts:
```diff
@@ -72,8 +74,15 @@ export function ɵɵcomponentHostSyntheticListener(
+  // this is where animations are defined.
+  // Note: this logic exists at runtime (not at compile time), since a Component can inherit a
+  // Directive (hence inherit its host listeners) and it would require loading component's renderer.
+  const renderer = currentDef === null || isComponentDef(currentDef) ?
```
this logic is same as in `ɵɵupdateSyntheticHostBinding`. Perhaps we could just move that into `loadComponentRenderer` directly?

pkozlowski-opensource commented:
@AndrewKushnir I can confirm that after the rebase the perf results from master and your PR are about the same so it is all good 👍 

I will do another review pass after you address Misko's review.

mhevery commented on packages/core/src/render3/instructions/host_property.ts:
```diff
@@ -70,7 +70,8 @@ export function ɵɵupdateSyntheticHostBinding<T>(
     const tNode = getSelectedTNode();
-    const renderer = loadComponentRenderer(tNode, lView);
+    const currentDef = getCurrentDirectiveDef(tView.data);
+    const renderer = loadComponentRenderer(currentDef, tNode, lView);
```
I wonder if we could do this a different way, can we discuss this over VC.

AndrewKushnir commented:
@mhevery @matsko thanks for the help and additional information. I've updated this PR to include extra comments to describe the `currentDef === null` check (due to root component creation process). Could you please have a look at this PR when you get a chance? Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:304101930:BASE:304101938:1585711363554:fdc07289)

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1928,9 +1931,12 @@ function getTViewCleanup(tView: TView): any[] {
-  return componentLView[RENDERER];
+export function loadComponentRenderer(
+    currentDef: DirectiveDef<any>| null, tNode: TNode, lView: LView): Renderer3 {
+  if (currentDef === null || isComponentDef(currentDef)) {
```
@mhevery there is a comment above this check, please let me know if you want me to include additional info or rephrase it. Thank you.

AndrewKushnir commented on integration/_payload-limits.json:
```diff
@@ -49,7 +49,7 @@
       "uncompressed": {
         "runtime-es2015": 2289,
-        "main-es2015": 226144,
+        "main-es2015": 226686,
```
This looks like an accumulated size increase, since the same happened in a [different PR (#36381)](https://github.com/angular/angular/pull/36381/files#diff-55315ab5c9e23bee7dc42a45a3517105R52).

kara commented:
@AndrewKushnir Can you update payload limits?

AndrewKushnir commented:
@kara thanks for the note. I've just updated payload limits: the payload size limit in integration tests in the `master` branch was higher that the one that was set in this PR (so I reverted my change) :) However there is still a payload limit increase for AIO, which is an accumulated increase from previous changes. I've also added a second commit with clang-related formatting changes. Thank you.

matsko commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1260,11 +1260,13 @@ function invokeDirectivesHostBindings(tView: TView, lView: LView, tNode: TNode)
-      const def = tView.data[i] as DirectiveDef<any>;
-      const directive = lView[i];
+    for (let dirIndex = start; dirIndex < end; dirIndex++) {
+      const def = tView.data[dirIndex] as DirectiveDef<any>;
```
could we get away with using `DirectiveRef<unkown>` here? That way the behavior of `any` won't seep into the framework.

AndrewKushnir replied:
I've used `DirectiveRef<unknown>` here as you proposed.
The `DirectiveRef<any>` is used more widely in codebase, but I guess that's mainly for historical reasons.

matsko commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1273,6 +1275,7 @@ function invokeDirectivesHostBindings(tView: TView, lView: LView, tNode: TNode)
     }
   } finally {
     setSelectedIndex(-1);
+    setCurrentDirectiveIndex(currentDirectiveIndex);
```
Is there a reason why need this?

matsko replied:
Is it because we no longer rely on the state being cleaned up once the `LFrame` is out of use?

AndrewKushnir replied:
The goal is to keep the global `state` in a clean state once we are done with host bindings processing (since there are other operations that happen afterwards while we process the view).

matsko commented on packages/core/src/render3/state.ts:
```diff
@@ -356,6 +357,21 @@ export function getCurrentDirectiveIndex(): number {
   return instructionState.lFrame.currentDirectiveIndex;
 }
 
+export function setCurrentDirectiveIndex(currentDirectiveIndex: number): void {
```
Can you add some documentation for this?

AndrewKushnir committed:
fixup! fix(core): handle synthetic props in Directive host bindings correctly

AndrewKushnir commented:
[Presubmit](http://test/OCL:306702047:BASE:306702106:1586980511193:fc2bc634).

kara commented on packages/core/test/acceptance/host_binding_spec.ts:
```diff
@@ -175,6 +178,274 @@ describe('host bindings', () => {
+         expect(events).toEqual(['@myAnimation.start', '@myAnimation.done']);
+       });
+  });
+
```
I don't see a test for a component with a directive on the same node - looks like https://github.com/angular/angular/pull/35568#discussion_r382349635 still needs to be addressed?

kara reviewed (approved):
LGTM once test is added

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/ui#id=OCL:306702047:BASE:307525392:1587436783212:38d20375)

AndrewKushnir committed:
fixup! fix(core): handle synthetic props in Directive host bindings correctly

AndrewKushnir commented:
[Presubmit](http://test/OCL:308670981:BASE:308671028:1588014546093:bf8fee70)
